### PR TITLE
feat(daemon): add Prime Agent RPC session resume

### PR DESCRIPTION
**Pull request by iamwaqargulzar**
## Summary

- add Prime Agent as a first-class runtime over the existing pi-compatible RPC bridge
- discover Prime's native JSONL sessions under `~/.prime/agent/sessions`
- resume the same native file with `prime-agent --mode rpc --resume <path>` instead of forking via `new_session(parentSession)`
- persist a uniquely discovered session path immediately while the run is live and retain it across canceled/failed terminal runs
- launch each fresh Prime child with a durable run-specific `--session-dir`, eliminating cross-process file attribution in Prime's shared directory
- preflight saved Prime session files before prompt trimming; a missing file clears the stale handle and re-seeds a fresh session with the full client transcript
- expose Prime in agent ordering, Settings, handoff targets, and bounded analytics

## Problem

Follow-up to #7261. Prime Agent 0.8.0 exposed two continuity gaps in the generic pi-RPC path:

1. `attachPiRpcSession()` only resolved the changed session file at `agent_end`, while `server.ts` persisted it only during terminal handling. Cancel/Retry or an application restart could therefore leave no `(conversationId, agentId)` session mapping and open another top-level Prime session.
2. `new_session(parentSession)` preserves history by creating a forked continuation. Prime supports true same-file RPC resume through its process-level `--resume <session-file>` option, which avoids multiplying top-level sessions.

Prime 0.8.0 silently creates a new file when `--resume` points to a missing path instead of returning an error. The daemon therefore validates this local file handle before composing the prompt; stale handles enter the existing fresh-session/full-transcript path before history can be trimmed.

The capture resolver remains conservative: it stores a handle only when exactly one JSONL file changed, preventing cross-conversation attribution when concurrent processes write the shared directory.

Fresh Prime children no longer scan that shared directory at all: Open Design assigns `~/.prime/agent/sessions/open-design/<run-id>` before spawn and passes it to both Prime and the capture bridge. Resume turns use the directory containing their already-attributed session file. This makes live discovery child-specific by construction while keeping the session durable across daemon restarts.

The activity-card `input.code` rendering tweak discussed in #7261 is intentionally excluded from this PR.

## Surface area checklist

- [x] UI: Prime appears in Settings and handoff targets; existing web visual coverage exercises the dynamic agent row.
- [x] CLI / environment: `PRIME_AGENT_BIN` is accepted as the per-agent executable override, with app-config coverage.
- [x] API / contract: Prime is included in the bounded analytics agent target union; no endpoint shape changes.
- [x] Default behavior: installed Prime users gain same-session RPC continuity and stale-file recovery; other runtimes keep their existing behavior.
- [ ] None of the above.

## Bug-fix verification

- `apps/daemon/tests/opencode-session-resume.test.ts`: server-level red/green coverage deletes Prime's captured session file between two turns and proves turn two starts without `--resume`, emits `agent_resume_auto_reseed`, and receives both the prior and current user transcript.
- `apps/daemon/tests/pi-rpc.test.ts`: proves the uniquely discovered session path is published during the live run rather than only at terminal finalization.
- `apps/daemon/tests/pi-rpc.test.ts`: also covers the requested race ordering—an unrelated shared-directory file changes before the child's first frame, the child's own file changes afterward, and only the isolated child path is published.
- `apps/daemon/tests/pi-rpc.test.ts`: covers the same ordering without a session-directory override for ordinary Pi; live capture remains disabled and terminal resolution rejects the two-file ambiguity.
- `apps/daemon/tests/app-config.test.ts`: proves `PRIME_AGENT_BIN` survives allowlist normalization.

## Validation

- `pnpm --filter @open-design/daemon typecheck`
- `pnpm --filter @open-design/web typecheck`
- 141 focused tests across `tests/pi-rpc.test.ts`, `tests/runtimes/prime-agent.test.ts`, `tests/opencode-session-resume.test.ts`, and `tests/app-config.test.ts`
- `pnpm guard`
- `git diff --check`
- manually verified Prime Agent 0.8.0 accepts `--mode rpc --resume` using a temporary copy of a real session file
- manually verified Prime Agent 0.8.0 silently creates a missing `--resume` path, motivating the preflight check

## Related

- Follow-up to #7261
- Related transport prerequisite: #7263

### CODE REVIEW

**Comment by lefarcen** [comment]
Thanks @iamwaqargulzar — the split from #7263 makes sense, and the Prime same-file `--resume` direction is clear from the write-up.

One small PR-body follow-up before review: could you add the Surface area checklist and tick the relevant boxes? This one touches daemon resume behavior plus the Settings / handoff surfaces, so that extra scoping will help reviewers see the user-facing paths quickly.

🧪 This PR also needs a manual QA pass before merge, so please hold off self-merging for now; we'll loop QA in once it's merge-ready.

> 💡 To drive this PR to merge hands-free, paste this to your AI coding agent (Claude Code / Codex / opencode / Cursor …):
> `Take over nexu-io/open-design#7265 until it merges — read https://raw.githubusercontent.com/nexu-io/looper/main/skills/pr-takeover/SKILL.md and follow it.`

File: apps/daemon/src/runtimes/defs/prime-agent.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    { id: 'low', label: 'Low' },
+    { id: 'medium', label: 'Medium' },
+    { id: 'high', label: 'High' },
+    { id: 'xhigh', label: 'Extra high' },
+    { id: 'max', label: 'Maximum' },
+  ],
+  buildArgs: (_prompt, _images, _dirs, options = {}, runtimeContext = {}) => {
+    const args = ['--mode', 'rpc'];
+    if (runtimeContext.cwd) args.push('--cwd', runtimeContext.cwd);
+    if (runtimeContext.resumeSessionId) args.push('--resume', runtimeContext.resumeSessionId);
```
Comment (PerishCode) [inline apps/daemon/src/runtimes/defs/prime-agent.ts:51]: Handle a missing `--resume` target instead of retrying the dead path forever. Prime resumes through process arguments here, but the child-close auto-reseed gate only admits `runtimeResumesSessionById(def)` or `resumesSessionViaAcpLoad`; `primeAgentDef` sets neither. It also has no Prime-specific branch in `isAgentResumeFailure`. If this JSONL file is deleted, pruned, or moved, Prime exits before RPC startup, the stored row remains, and every later turn skips the transcript and launches the same invalid `--resume` again. Extend the resume capability/failure policy to cover this process-argument mode, recognize Prime's missing-session failure signal, clear the stale mapping, and transparently rerun with the full transcript. Add a server-level regression test that seeds a missing Prime path and verifies one fresh reseed rather than a permanent failure loop.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/agent-protocol/pi-rpc/session.ts [hunk 2]
```diff
@@ -331,6 +344,10 @@ export function attachPiRpcSession({
     // acting on the parsed objects.
     if (finished) return;
 
+    // Capture while the run is live so cancellation, failure, and application
+    // restart do not discard the native session handle.
+    captureSessionPath();
```
Comment (PerishCode) [inline apps/daemon/src/agent-protocol/pi-rpc/session.ts]: Persist the discovered handle while the run is live if application-restart continuity is part of this fix. `captureSessionPath()` only assigns `capturedSessionPath` in this in-memory session object; the database write still occurs later in `server.ts` during child finalization through `getLastSessionPath()`. A daemon/application restart between this callback and finalization loses both the object and the mapping, so the next launch creates another top-level Prime session despite the changed comment and PR problem statement. Provide a capture callback (with the conversation/model/cwd/cursor metadata owned by `server.ts`) that durably upserts the uniquely identified path as soon as it is discovered, or narrow the promised behavior and add a test for the actual restart boundary.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by PerishCode — CHANGES_REQUESTED** [review]
Two session-lifecycle gaps still break the continuity guarantees of this change: stale Prime resume targets never enter the existing auto-reseed path, and live path discovery is not durable across a daemon restart. I reviewed all 12 changed files and the surrounding resume/finalization control flow. Local focused tests could not be rerun because this prepared worktree has no installed dependencies.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Thanks for the quick iteration here. @PerishCode's current-head review already covers the two blockers that matter most on this patch: stale `--resume` targets need to fall back into the existing reseed path, and the discovered Prime session handle needs to be made durable across a daemon restart if restart continuity is part of the contract.

Once those two points are addressed on this head, we can take another pass from there.

File: apps/daemon/src/runtimes/defs/prime-agent.ts (hunk 1 above)
Comment (iamwaqargulzar) [reply apps/daemon/src/runtimes/defs/prime-agent.ts:51]: Addressed in a385f2cb0. Prime 0.8.0 does not surface an error for a missing `--resume` target; it silently creates that path, so post-exit error matching cannot recover the lost context. The daemon now checks the saved local session file before prompt composition, clears a stale handle, emits `agent_resume_auto_reseed`, and continues through the fresh/full-transcript path. The new server-level regression test deletes the captured file between turns and verifies no `--resume` argv plus both prior and current user transcript in turn two.

File: apps/daemon/src/agent-protocol/pi-rpc/session.ts (hunk 2 above)
Comment (iamwaqargulzar) [reply apps/daemon/src/agent-protocol/pi-rpc/session.ts]: Addressed in a385f2cb0. `attachPiRpcSession` now publishes the uniquely discovered path through `onSessionPath` as soon as discovery succeeds, and `server.ts` immediately upserts the session mapping with the current stable-prompt metadata/cursor. Terminal persistence remains as a fallback. The pi-RPC test now asserts that the live callback fires exactly once with the captured path.

File: apps/daemon/src/agent-protocol/pi-rpc/session.ts [hunk 3]
```diff
@@ -331,6 +349,10 @@ export function attachPiRpcSession({
...
```
Comment (PerishCode) [inline apps/daemon/src/agent-protocol/pi-rpc/session.ts]: Do not latch the first transient sole change in the shared Prime session directory. `captureSessionPath()` now runs for every JSON frame and permanently returns early after the first match. If another Prime process updates file B after this run's snapshot but before this child has written file A, the first frame from this child sees exactly one changed file (B), stores B through `onSessionPath`, and later ignores A. The next turn then resumes another conversation's native history, which is precisely the cross-conversation attribution the one-file rule is intended to prevent. Derive the path from an event/session identifier attributable to this child, or otherwise delay/validate the candidate until there is child-specific evidence before persisting it; do not make a transient directory-wide observation irreversible. Add a fixture-matrix case where a foreign file changes before this child's first frame and this child's own file changes afterward, asserting that the foreign path is never published.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by PerishCode — CHANGES_REQUESTED** [review]
The stale-handle preflight and immediate persistence address the two previously reported lifecycle gaps, but the live capture introduces a cross-conversation attribution race in Prime's shared session directory. This needs to be resolved before merge.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by xxiaoxiong — APPROVED** [review]

**Comment by lefarcen** [comment]
Thanks for the update — @PerishCode's latest current-head review narrows this to one remaining blocker: the live session-path capture now needs to avoid latching a foreign file from Prime's shared session directory before this run has child-specific evidence.

Once that race is closed on this head, the remaining review path should be much clearer.

File: apps/daemon/src/agent-protocol/pi-rpc/session.ts (hunk 3 above)
Comment (iamwaqargulzar) [reply apps/daemon/src/agent-protocol/pi-rpc/session.ts]: Addressed in 73cc56730 by removing the shared-directory attribution problem before spawn. Every fresh Prime child now receives a durable run-specific `--session-dir` (`<Prime sessions>/open-design/<run-id>`), and `attachPiRpcSession` scans only that directory. Resume turns use `dirname(resumeSessionId)`, so the already-attributed file remains durable and reopenable after daemon restart. Added the requested ordering fixture: a foreign file changes before the child first frame, no path is published, then the child file changes and only that isolated path is published. The server lifecycle test also asserts Prime receives an `open-design/<run-id>` session path. Validation: 140 focused tests, daemon/web typechecks, guard, and diff check all pass.

Comment (PerishCode) [inline apps/daemon/src/agent-protocol/pi-rpc/session.ts]: Keep live capture from latching a foreign session for ordinary Pi. This callback now runs on the first parsed frame and `capturedSessionPath` makes the first sole changed file permanent. Prime is protected because `server.ts` supplies its run-specific `sessionDir`, but the existing `pi` definition supplies no override, so it still scans `<cwd>/.pi/sessions`. If another Pi process changes file B before this child emits its first frame, B is persisted immediately; when this child later writes A, the early-return prevents correction and the next turn loads the wrong conversation. Preserve live capture only when the directory is child-specific, or add an equivalent per-run attribution mechanism for plain Pi; otherwise retain terminal-time conservative resolution for shared directories. Add a matrix case with no `sessionDir` where a foreign file changes before the first frame and the child's file changes later, asserting the foreign path is never published.

<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Review by PerishCode — CHANGES_REQUESTED** [review]
The Prime-specific per-run directory fixes the previously reported attribution race for Prime, but the live-capture change also applies to ordinary Pi, which still scans its shared workspace session directory. That regression can persist another Pi process's session and must be fixed before merge.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Thanks for pushing on this quickly — @PerishCode's latest current-head review narrows it again to one remaining blocker: the new live-capture path still needs to avoid persisting a foreign session for ordinary Pi, not just Prime.

Once that shared-directory case is handled on this head, the remaining review path should be much cleaner.

**Review by PerishCode — APPROVED** [review]
@iamwaqargulzar I reviewed the full current-head diff, including Prime's process-argument resume path, stale-file preflight and full-transcript reseed, live durable capture, per-run session-directory isolation, ordinary Pi's conservative shared-directory behavior, runtime registration/config exposure, UI handoff surfaces, analytics typing, and the focused regression coverage. The earlier lifecycle and attribution races are resolved on this head, and I found no remaining actionable issue in the changed ranges. Nice work iterating carefully through the concurrency and restart boundaries here.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/daemon/src/agent-protocol/pi-rpc/session.ts (hunk 3 above)
Comment (iamwaqargulzar) [reply apps/daemon/src/agent-protocol/pi-rpc/session.ts]: Addressed in 7c5a41a18. Live capture is now explicitly opt-in through `onSessionPath`, and `server.ts` supplies that callback only for process-argument resume runtimes with the child-specific session directory (Prime). Ordinary Pi receives no live callback: parsed frames, failures, and aborts cannot latch a shared-directory candidate; successful `agent_end` retains the original terminal-time conservative resolver. Added the requested no-`sessionDir` matrix case: foreign B changes before the first frame, remains uncaptured, own A changes later, and terminal resolution returns null for the two-file ambiguity. Validation passes: 141 focused tests, daemon/web typechecks, guard, and diff check.

**Comment by iamwaqargulzar** [comment]
Rebased onto current `main` (`d488b1aa8`) and resolved the `apps/daemon/src/server.ts` conflicts by retaining the latest upstream prompt-image/projection-root flow together with Prime's isolated per-run session directory and durable live capture.

Validation on revised head `50626855c`:
- focused Prime/pi-RPC/resume/config suite: 150 passed
- daemon typecheck: passed
- web typecheck: passed
- `pnpm guard`: passed
- `git diff --check`: passed
- real Prime Agent 0.8.0 resume smoke: `--resume` reopened the exact copied JSONL session and `get_state` succeeded over RPC stdin

The branch is ready for renewed review and the requested manual QA pass.

**Review by PerishCode — APPROVED** [review]
@iamwaqargulzar I reviewed every changed range on the rebased head, including Prime's same-file process-argument resume, stale-file full-transcript reseeding, child-specific session-directory attribution, live durable capture, ordinary Pi's conservative shared-directory behavior, runtime/config/UI registration, analytics typing, and the focused regression coverage. The previous lifecycle and attribution blockers remain correctly resolved after the rebase, and the Prime 0.8.0 CLI source confirms the argument and session-directory contracts used here. Nice work carrying this through several subtle concurrency and restart-continuity edge cases.

Local test execution was unavailable because this prepared reviewer worktree has no installed dependencies; the live PR checks were still pending at review time.


<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=codex · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Thanks @xiaoche-hub — that screenshot / recording set is the right bar for this one.

@iamwaqargulzar, once the rebased head includes those captures, it should make the Prime selection, handoff target, and same-session resume path much easier to verify end to end.
